providers/default.rb

```diff
@@ -49,99 +49,273 @@ def whyrun_supported?
+    else
+      puts "No key found at #{node[:encrypted_blockdevice][:encrypted_data_bag_secret]}"
+    end
+
+    # If we can't find the item for the device we're creating, the results should be empty or nil.
+    if ( keystore_item_result == nil || keystore_item_result.empty? )
+
+      # This is probably the first run of this cookbook on this node with this configured device.
+      # So we set about creating a key, opening the device with the details we have and then saving the details to the keystore. 
+      
+      # This is a new device - i.e. one for which we cannot find existing data, so we generate a new key.
+      key = `openssl rand -hex #{@new_resource.keylength} | tr -d '\r\n'`
+     
+      puts "Generating #{@new_resource.keylength} byte key for #{name}"
+ 
+      # We determine our 'keyfilesize' for --keyfile-size based on the 'keylength' in bytes doubled because the key is hex-encoded.
+      keyfilesize = (new_resource.keylength * 2)
+
+      # We pass the key without a newline to the cryptsetup command with the necessary arguments.
+      # We should find a way to pipe this in without it showing it in ps auxww or any logs.
+
+      `echo -n #{key} | cryptsetup #{cryptsetup_args} #{device_name_order} --keyfile-size #{keyfilesize} --cipher #{cipher} --batch-mode --key-file=-`
+
+      puts `cryptsetup status #{name}`
+       
+      # We probably created the device, so we go on to create the bag item.
+      puts "Creating new device item for #{name}" 
+
+      # We flesh out the databag of the used settings and key for the keystore - rather useful after a reboot.
+      new_deviceitem = {
+        "id" => keystore_item_name,
+        "name" => name,
+        "device" => device,
+        "cryptsetup_args" => cryptsetup_args,
+        "keyfilesize" => keyfilesize,
+        "cipher" => cipher,
+        "key" => key
+      }
+
+      # Since we have two modes of databag storage, we have a minor divergence in behaviour - both save the settings/key to the keystore. 
+      if @new_resource.keystore == "encrypted_databag"
+        # Encrypted databag item.
+        # The EncryptedDataBagItem object does not allow direct .save functionality because Opscode seem to think encrypted things ought to be readonly by machines.
+        # So this is a workaround based upon their own test spec that demonstrates this behaviour. See Chef repo: spec/unit/knife/data_bag_create_spec.rb
+        # Go complain at https://tickets.opscode.com/browse/CHEF-2401
+        puts "Encrypting device item for #{name}"
+        encrypted_new_deviceitem = Chef::EncryptedDataBagItem.encrypt_data_bag_item(new_deviceitem, encrypted_data_bag_secret)
+        deviceitem = Chef::DataBagItem.from_hash(encrypted_new_deviceitem)
+      else
+        # Unencrypted databag item.
+        deviceitem = Chef::DataBagItem.new
+        deviceitem.raw_data = new_deviceitem
+      end 
+      deviceitem.data_bag(keystore_databag_name)
+      deviceitem.save
```

Comment:
This bit is 403ing when the databag exists and the clients have update perms on that databag.